// @yarnpkg/pnpify/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PortableNodeModulesFS = exports.NodeModulesFS = exports.resolveNodeModulesPath = exports.patchFs = exports.SdkCommand = exports.RunCommand = void 0;
const tslib_1 = require("tslib");
const RunCommand_1 = tslib_1.__importDefault(require("./commands/RunCommand"));
exports.RunCommand = RunCommand_1.default;
const SdkCommand_1 = tslib_1.__importDefault(require("./commands/SdkCommand"));
exports.SdkCommand = SdkCommand_1.default;
var index_1 = require("./index");
Object.defineProperty(exports, "patchFs", { enumerable: true, get: function () { return index_1.patchFs; } });
var resolveNodeModulesPath_1 = require("./resolveNodeModulesPath");
Object.defineProperty(exports, "resolveNodeModulesPath", { enumerable: true, get: function () { return resolveNodeModulesPath_1.resolveNodeModulesPath; } });
var NodeModulesFS_1 = require("./NodeModulesFS");
Object.defineProperty(exports, "NodeModulesFS", { enumerable: true, get: function () { return NodeModulesFS_1.NodeModulesFS; } });
Object.defineProperty(exports, "PortableNodeModulesFS", { enumerable: true, get: function () { return NodeModulesFS_1.PortableNodeModulesFS; } });
